contracts/transmuter/src/transmuter_pool/asset_group.rs

```diff
@@ -0,0 +1,277 @@
+use std::collections::BTreeMap;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{ensure, Decimal, Uint64};
+
+use crate::{corruptable::Corruptable, transmuter_pool::MAX_ASSET_GROUPS, ContractError};
+
+use super::TransmuterPool;
+
+#[cw_serde]
+pub struct AssetGroup {
+    denoms: Vec<String>,
+    is_corrupted: bool,
+}
+
+impl AssetGroup {
+    pub fn new(denoms: Vec<String>) -> Self {
+        Self {
+            denoms,
+            is_corrupted: false,
+        }
+    }
+
+    pub fn denoms(&self) -> &[String] {
+        &self.denoms
+    }
+
+    pub fn into_denoms(self) -> Vec<String> {
+        self.denoms
+    }
+
+    pub fn add_denoms(&mut self, denoms: Vec<String>) -> &mut Self {
+        self.denoms.extend(denoms);
+        self
+    }
+
+    pub fn remove_denoms(&mut self, denoms: Vec<String>) -> &mut Self {
+        self.denoms.retain(|d| !denoms.contains(d));
+        self
+    }
+}
+
+impl Corruptable for AssetGroup {
+    fn is_corrupted(&self) -> bool {
+        self.is_corrupted
+    }
+
+    fn mark_as_corrupted(&mut self) -> &mut Self {
+        self.is_corrupted = true;
+        self
+    }
+
+    fn unmark_as_corrupted(&mut self) -> &mut Self {
+        self.is_corrupted = false;
+        self
+    }
+}
+
+impl TransmuterPool {
+    pub fn has_asset_group(&self, label: &str) -> bool {
+        self.asset_groups.contains_key(label)
+    }
+
+    pub fn mark_corrupted_asset_group(&mut self, label: &str) -> Result<&mut Self, ContractError> {
+        self.asset_groups
+            .get_mut(label)
+            .ok_or_else(|| ContractError::AssetGroupNotFound {
+                label: label.to_string(),
+            })?
+            .mark_as_corrupted();
+
+        Ok(self)
+    }
+
+    pub fn unmark_corrupted_asset_group(
+        &mut self,
+        label: &str,
+    ) -> Result<&mut Self, ContractError> {
+        self.asset_groups
+            .get_mut(label)
+            .ok_or_else(|| ContractError::AssetGroupNotFound {
+                label: label.to_string(),
+            })?
+            .unmark_as_corrupted();
+
+        Ok(self)
+    }
+
+    pub fn create_asset_group(
+        &mut self,
+        label: String,
+        denoms: Vec<String>,
+    ) -> Result<&mut Self, ContractError> {
+        // ensure that asset group does not already exist
+        ensure!(
+            !self.asset_groups.contains_key(&label),
+            ContractError::AssetGroupAlreadyExists {
+                label: label.clone()
+            }
+        );
+
+        // ensure that all denoms are valid pool assets
+        for denom in &denoms {
+            ensure!(
```

Comment:
Is there a DOS vector here if I add an asset group with the same asset a ton of times? Might be worth enforcing uniqueness of elements